File: backend/app/routes/blob.py
from fastapi import APIRouter, Request

from app.utils.storage import blob_enabled, generate_client_token

router = APIRouter()


@router.get("/status")
def blob_status():
    return {"blobEnabled": blob_enabled()}


@router.post("/token")
async def blob_token(request: Request):
    if not blob_enabled():
        return {"blobEnabled": False, "mode": "local"}

    body = await request.json()
    event_type = body.get("type")

    if event_type == "blob.generate-client-token":
        payload = body.get("payload") or {}
        pathname = payload.get("pathname") or "uploads/document.pdf"
        return {
            "type": "blob.generate-client-token",
            "clientToken": generate_client_token(pathname),
        }

    if event_type == "blob.upload-completed":
        return {"type": "blob.upload-completed"}

    pathname = body.get("filename") or body.get("pathname") or "uploads/document.pdf"
    return {
        "type": "blob.generate-client-token",
        "clientToken": generate_client_token(pathname),
        "blobEnabled": True,
    }

File: frontend/src/services/blob.js
let blobModeCache = null

export async function getStorageMode() {
  if (blobModeCache) return blobModeCache
  try {
    const response = await fetch('/api/blob/status')
    const data = await response.json()
    blobModeCache = data?.blobEnabled ? 'blob' : 'local'
  } catch {
    blobModeCache = 'local'
  }
  return blobModeCache
}

export async function uploadPdfs(files) {
  const mode = await getStorageMode()
  if (mode !== 'blob') {
    return { mode: 'local', files: Array.from(files) }
  }

  const { upload } = await import('@vercel/blob/client')
  const uploaded = []
  for (const file of files) {
    const blob = await upload(`uploads/${file.name}`, file, {
      access: 'public',
      handleUploadUrl: '/api/blob/token',
    })
    uploaded.push({ url: blob.url, name: file.name })
  }
  return { mode: 'blob', files: uploaded }
}
